src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs

```diff
@@ -469,8 +469,46 @@ public bool TryGetDateTime(out DateTime value)
                 throw ThrowHelper.GetInvalidOperationException_ExpectedString(TokenType);
             }
 
-            ReadOnlySpan<byte> span = HasValueSequence ? ValueSequence.ToArray() : ValueSpan;
-            return JsonHelpers.TryParseAsISO(span, out value, out int bytesConsumed) && span.Length == bytesConsumed;
+            ReadOnlySpan<byte> span = stackalloc byte[0];
+
+            if (HasValueSequence)
+            {
+                long sequenceLength = ValueSequence.Length;
+
+                if (!JsonReaderHelper.IsValidDateTimeOffsetParseLength(sequenceLength))
+                {
+                    value = default;
+                    return false;
+                }
+
+                Debug.Assert(sequenceLength <= JsonConstants.MaximumEscapedDateTimeOffsetParseLength);
+                Span<byte> stackSpan = stackalloc byte[(int)sequenceLength];
+
+                ValueSequence.CopyTo(stackSpan);
+                span = stackSpan;
+            }
+            else
+            {
+                if (!JsonReaderHelper.IsValidDateTimeOffsetParseLength(ValueSpan.Length))
```

Comment:
nit: Maybe we should add an int32 overload of `IsValidDateTimeOffsetParseLength` to avoid forcing the more common int32 code path to be extended to int64.